src/proxy/metadata.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::convert::Infallible;
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+
+use std::time::Duration;
+
+use anyhow::anyhow;
+use bytes::Bytes;
+use http_body_util::Full;
+use hyper::service::service_fn;
+use hyper::{Response, StatusCode};
+
+use tokio::net::TcpStream;
+use tracing::{error, trace};
+
+use crate::proxy::connection_manager::{ConnectionManager, ConnectionTuple};
+
+use super::Error;
+
+/// METADATA_SERVER_IP provides the well-known metadata server IP.
+/// This is captured by the redirection.
+pub const METADATA_SERVER_IP: IpAddr = IpAddr::V4(Ipv4Addr::new(169, 254, 169, 111));
```

Comment:
Probably need to document this like we do for the other link-local addresses we reserve.

Reply:
Ack. An alternative is to not do anything special and make a real service like https://github.com/istio/istio/pull/45009.. not sure what is preferred

Reply:
We could also bind another in-pod port which would make it explicitly private and effectively remove the need to do identity checks or reserve an IP or expose this in the cluster - but I don't care much.

I am fine with link-local for now if we toss it into `ARCHITECTURE.md`

Reply:
Lets do the link local thing for now but ill make sure its documented + explicitly called out as experimental. This feature will need a lot of feedback overall